retro/reco.py

```diff
@@ -425,6 +507,507 @@ def loglike(cube, ndim=None, nparams=None): # pylint: disable=unused-argument
+                               ('z', np.float32),
+                               ('sinzen', np.float32),
+                               ('coszen', np.float32),
+                               ('sinaz', np.float32),
+                               ('cosaz', np.float32),
+                               ])
+
+        def create_x(x_cart, x_spher):
+            '''
+            patch back together the cartesian and spherical coordinates into one array
+            '''
+            # ToDO: make proper
+            x = np.empty(n)
+            x[:n_cart] = x_cart
+            x[n_cart+1::2] = x_spher['zen']
+            x[n_cart::2] = x_spher['az']
+            return x
+
+        def fill_from_spher(s):
+            '''
+            fill in the remaining values giving the two angles `zen` and `az`
+            '''
+            s['sinzen'] = np.sin(s['zen'])
+            s['coszen'] = np.cos(s['zen'])
+            s['sinaz'] = np.sin(s['az'])
+            s['cosaz'] = np.cos(s['az'])
+            s['x'] = s['sinzen'] * s['cosaz']
+            s['y'] = s['sinzen'] * s['sinaz']
+            s['z'] = s['coszen']
+
+        def fill_from_cart(s_vector):
+            '''
+            fill in the remaining values giving the cart, coords. `x`, `y` and `z`
+            '''
+
+            for s in s_vector:
+                radius = np.sqrt(s['x']**2 + s['y']**2 + s['z']**2)
+                if not radius == 0:
+                    # make sure they're length 1
+                    s['x'] /= radius
+                    s['y'] /= radius
+                    s['z'] /= radius
+                    s['az'] = np.arctan2(s['y'], s['x']) % (2 * np.pi)
+                    s['coszen'] = s['z']
+                    s['zen'] = np.arccos(s['coszen'])
+                    s['sinzen'] = np.sin(s['zen'])
+                    s['sinaz'] = np.sin(s['az'])
+                    s['cosaz'] = np.cos(s['az'])
+                else:
+                    s['z'] = 1
+                    s['az'] = 0
+                    s['zen'] = 0
+                    s['coszen'] = 1
+                    s['sinzen'] = 0
+                    s['cosaz'] = 1
+                    s['sinaz'] = 0
+
+
+        def reflect(old, centroid, new):
+            '''
+            reflect the old point around the centroid into the new point on the sphere
+            '''
+
+            x = old['x']
+            y = old['y']
+            z = old['z']
+
+            ca = centroid['cosaz']
+            sa = centroid['sinaz']
+            cz = centroid['coszen']
+            sz = centroid['sinzen']
+            
+            new['x'] = 2*ca*cz*sz*z + x*(ca*(-ca*cz**2 + ca*sz**2) - sa**2) + y*(ca*sa + sa*(-ca*cz**2 + ca*sz**2))
+            new['y'] = 2*cz*sa*sz*z + x*(ca*sa + ca*(-cz**2*sa + sa*sz**2)) + y*(-ca**2 + sa*(-cz**2*sa + sa*sz**2))
+            new['z'] = 2*ca*cz*sz*x + 2*cz*sa*sz*y + z*(cz**2 - sz**2)
+
+            fill_from_cart(new)
+
+        #N = 10 * (n + 1)
```

Comment:
looks like your indentation is weird? did you use tabs/spaces/?

Reply:
don't understand..?